### DISCUSSION TITLE
Set exposure bias?

### DISCUSSION THREAD

**Discussion by abey79** [raspberrypi/picamera2#131] [Q&A] [answered]
I would like to add the ability to adjust exposure bias to the still capture example, but I couldn't find how this is done. Is there some example for this somewhere?

**Comment by abey79** [comment]
I figured part of it: `campi2.set_controls({'ExposureValue': 2})`. However I find that I need to capture 2-3 stills until the exposure is correctly set. Is there a way to make it more predictable?

**Comment by kbingham** [comment]
Currently libcamera doesn't fullfil a guarantee for 'per frame controls' ... this is because it can take up to 3 frames to set the Exposure, so we have to rework the internals to ensure that the exposure is set 3 frames earlier than requested.
It's actually something we met up for last week, and are actively working towards fixing, but it is complicated and may take a bit of time. The FrameMetadata should however tell you the correct frame that the exposure was finally set correctly after the sensor delays.

**Reply by abey79** [reply]
Thanks for the detailed explanation—it makes sense. I'm wondering if there is a way to "cheat" by having a preview stream going on (with `Preview.NULL` because I have a headless setup) and using a sequence such as:
```
picam2.set_controls({"ExposureValue": ev})
time.sleep(0.2)  # let the preview run for a few frames so that ExposureValue is taken into account
picam2. switch_mode_and_capture_array()
# ...
```

Would you expect that to work?

**Comment by davidplowman** [comment]
Could you say a little more about what you're trying to do? I'd expect changing the ev and having a short pause would work. But maybe you can just run the camera permanently with higher ev? Or possibly you could stop the camera and recalculate the exposure/gain for yourself before doing the capture. [This example](https://github.com/raspberrypi/picamera2/blob/main/examples/opencv_mertens_merge.py) does something a bit like that.

**Comment by abey79** [comment]
I'm trying to do the same as what I alluded to in my [other post](https://github.com/raspberrypi/picamera2/discussions/135#discussioncomment-2813541). The idea is a still image server were the parameters (in particular ev) can be passed in the HTTP request and directly applied to the returned image.

**Comment by davidplowman** [comment, marked as answer]
Just to add one more thing on this topic. The thing about "ev" is that it adjusts the target the AGC/AEC is aiming for, so once you change it, it takes a little while to have an effect. If you want to adjust the exposure just for a capture, this kind of thing works:
```
from picamera2 import Picamera2
import time

picam2 = Picamera2()
preview_config = picam2.preview_configuration()
still_config = picam2.still_configuration()
picam2.configure(preview_config)
picam2.start()

while True:
    time.sleep(2)
    metadata = picam2.capture_metadata()
    exposure = metadata["ExposureTime"]
    gain = metadata["AnalogueGain"] * metadata["DigitalGain"]
    picam2.stop()
    
    picam2.configure(still_config)
    picam2.set_controls({"ExposureTime": exposure*2, "AnalogueGain": gain*2})
    picam2.start()
    picam2.capture_file("test.jpg")
    picam2.stop()
    
    picam2.configure(preview_config)
    picam2.set_controls({"ExposureTime": exposure, "AnalogueGain": gain})
    picam2.start()
    picam2.set_controls({"AeEnable": True})
```
This script runs a preview and then, every 2 seconds, takes a full resolution capture. The capture has twice the exposure and twice the gain of the normally exposed image. The key point is that setting a new exposure time and gain _before_ restarting the camera means it takes effect immediately.

When we return to preview, notice how we set the exposure and gain back again before the camera is restarted, so be sure we go straight back to the normal exposure level without the images being briefly over-exposed. Afterwards we also re-enable the AGC/AEC again.

**Reply by abey79** [reply]
This looks great, it can certainly work for me.
